src/webapi/web_socket.rs

```diff
@@ -0,0 +1,197 @@
+        js!( return @{self}.extensions; ).try_into().unwrap()
+    }
+
+    /// Returns a string indicating the name of the sub-protocol the server selected;
+    /// this will be one of the strings specified in the protocols parameter when
+    /// creating the WebSocket object.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket)
+    pub fn protocol(&self) -> String {
+        js!( return @{self}.protocol; ).try_into().unwrap()
+    }
+
+    /// Returns a string indicating the name of the sub-protocol the server selected;
+    /// this will be one of the strings specified in the protocols parameter when
+    /// creating the WebSocket object.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket)
+    pub fn ready_state(&self) -> ReadyState {
+        js!( return @{self}.protocol; ).try_into().unwrap()
+    }
+
+    /// Returns the URL as resolved by the constructor. This is always an absolute URL.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket)
+    pub fn url(&self) -> String {
+        js!( return @{self}.url; ).try_into().unwrap()
+    }
+
+    /// Closes the WebSocket connection or connection attempt, if any. If the connection
+    /// is already CLOSED, this method does nothing.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket#close())
+    pub fn close(&self) {
+        js!( @(no_return) @{self}.close(); );
+    }
+
+    /// Closes the WebSocket connection or connection attempt, if any. If the connection
+    /// is already CLOSED, this method does nothing.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket#close())
+    pub fn close_with_status(&self, code: CloseEventCode, reason: &str) {
+        let code = code.0;
+        js!( @(no_return) @{self}.close(@{code}, @{reason}); );
+    }
+
+    /// Enqueues the specified data to be transmitted to the server over the WebSocket
+    /// connection, increasing the value of bufferedAmount by the number of bytes needed
+    /// to contain the data. If the data can't be sent (for example, because it needs to
+    /// be buffered but the buffer is full), the socket is closed automatically.
+    pub fn send_text(&self, text: &str) {
+        js!( @(no_return) @{self}.send(@{text}); );
+    }
+
+    fn _send_binary(&self, binary: Reference) {
+        js!( @(no_return) @{self}.send(@{binary}); );
+    }
+
+    /// Enqueues the specified data to be transmitted to the server over the WebSocket
+    /// connection, increasing the value of bufferedAmount by the number of bytes needed
+    /// to contain the data. If the data can't be sent (for example, because it needs to
+    /// be buffered but the buffer is full), the socket is closed automatically.
+    pub fn send_binary<T: AsBinaryRef>(&self, binary: &T) {
+        self._send_binary(binary.as_binary_ref());
+    }
```

Comment:
For now I *think* I'd prefer having separate methods for each type instead of using a trait to emulate overloading. (And we have a separate method for sending text anyway.)

Also, the way the arrays are currently converted:

`fn as_binary_ref(&self) -> Reference { TypedArray::from(self).as_ref().clone() }`

is somewhat inefficient; it's better to use `UnsafeTypedArray` which makes it zero-copy (see `send_with_bytes` in `xml_http_request.rs`)

Reply:
The reason I split it into just two methods (text/binary) is because this is how messages are mapped onto the underlying websocket protocol: the protocol distinguishes text from binary messages, but doesn't distinguish different types of binary message (hence the need for `binary_type`).

I think using separate methods was falsely give the impression that each type could be sent independently, and the receiving end would always receive the same type that was sent.

I will take a look at `UnsafeTypedArray`.

Reply:
> I think using separate methods was falsely give the impression that each type could be sent independently, and the receiving end would always receive the same type that was sent.

Well, we could document that, no? The original JavaScript API has only one method doesn't fare any better in the confusion department.

In general I think that in cases where this is not meant to be extensible by the user **and** the amount of types we have to support is very small and strictly limited then having separate methods is somewhat better. It plays better with type inference, works with auto-deref**, it's easier to use and discover (you don't have to check what impls the trait has), and you don't need to define an extra trait.

** Example:

```rust
use std::ops::Deref;

trait Foo {}

fn function_generic< T: Foo >( value: &T ) {}
fn function_concrete( value: &Bar ) {}

struct Bar;
impl Foo for Bar {}

struct MyType( Bar );
impl Deref for MyType {
    type Target = Bar;
    fn deref( &self ) -> &Self::Target { &self.0 }
}



fn main() {
    let value = MyType( Bar );
    function_concrete( &value ); // Works!
    function_generic( &value ); // the trait `Foo` is not implemented for `MyType`
}
```